connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/LiteralExpressionProtoConverter.scala

```diff
@@ -134,46 +192,46 @@ object LiteralExpressionProtoConverter {
       builder.result()
     }
 
-    val elementType = array.getElementType
-    if (elementType.hasShort) {
-      makeArrayData(v => v.getShort.toShort)
-    } else if (elementType.hasInteger) {
-      makeArrayData(v => v.getInteger)
-    } else if (elementType.hasLong) {
-      makeArrayData(v => v.getLong)
-    } else if (elementType.hasDouble) {
-      makeArrayData(v => v.getDouble)
-    } else if (elementType.hasByte) {
-      makeArrayData(v => v.getByte.toByte)
-    } else if (elementType.hasFloat) {
-      makeArrayData(v => v.getFloat)
-    } else if (elementType.hasBoolean) {
-      makeArrayData(v => v.getBoolean)
-    } else if (elementType.hasString) {
-      makeArrayData(v => v.getString)
-    } else if (elementType.hasBinary) {
-      makeArrayData(v => v.getBinary.toByteArray)
-    } else if (elementType.hasDate) {
-      makeArrayData(v => DateTimeUtils.toJavaDate(v.getDate))
-    } else if (elementType.hasTimestamp) {
-      makeArrayData(v => DateTimeUtils.toJavaTimestamp(v.getTimestamp))
-    } else if (elementType.hasTimestampNtz) {
-      makeArrayData(v => DateTimeUtils.microsToLocalDateTime(v.getTimestampNtz))
-    } else if (elementType.hasDayTimeInterval) {
-      makeArrayData(v => IntervalUtils.microsToDuration(v.getDayTimeInterval))
-    } else if (elementType.hasYearMonthInterval) {
-      makeArrayData(v => IntervalUtils.monthsToPeriod(v.getYearMonthInterval))
-    } else if (elementType.hasDecimal) {
-      makeArrayData(v => Decimal(v.getDecimal.getValue))
-    } else if (elementType.hasCalendarInterval) {
-      makeArrayData(v => {
-        val interval = v.getCalendarInterval
-        new CalendarInterval(interval.getMonths, interval.getDays, interval.getMicroseconds)
-      })
-    } else if (elementType.hasArray) {
-      makeArrayData(v => toArrayData(v.getArray))
-    } else {
-      throw InvalidPlanInput(s"Unsupported Literal Type: $elementType)")
+    makeArrayData(getConverter(array.getElementType))
+  }
+
+  private def toMapData(map: proto.Expression.Literal.Map): Any = {
+    def makeMapData[K, V](
+        keyConverter: proto.Expression.Literal => K,
+        valueConverter: proto.Expression.Literal => V)(implicit
+        tagK: ClassTag[K],
+        tagV: ClassTag[V]): mutable.Map[K, V] = {
+      val builder = mutable.HashMap.empty[K, V]
+      val keys = map.getKeysList.asScala
+      val values = map.getValuesList.asScala
+      builder.sizeHint(keys.size)
+      keys.zip(values).foreach { case (key, value) =>
+        builder += ((keyConverter(key), valueConverter(value)))
+      }
+      builder
+    }
+
+    makeMapData(getConverter(map.getKeyType), getConverter(map.getValueType))
+  }
+
+  private def toStructData(struct: proto.Expression.Literal.Struct): Any = {
+    val elements = struct.getElementsList.asScala
+    val dataTypes = struct.getStructType.getStruct.getFieldsList.asScala.map(_.getDataType)
+    val structData = elements
```

Comment:
Maybe we can define a method similar to the following:

```
    def toTuple[A <: Object](data: Seq[A]): Product = {
      try {
        val tupleClass = Utils.classForName("scala.Tuple" + data.length)
        tupleClass.getConstructors.head.newInstance(data: _*).asInstanceOf[Product]
      } catch {
        case _: Exception =>
          throw InvalidPlanInput(s"Unsupported Literal: " +
            s"${data.mkString("Array(", ", ", ")")})")
      }
    }
```

then 

```
    val structData = elements
      .zip(dataTypes)
      .map { case (element, dataType) =>
        getConverter(dataType)(element)
      }.toSeq.asInstanceOf[Seq[Object]]
    toTuple(structData)
```

Some optimizations can be made for constructors in `toTuple`, such as constructor method caching

Reply:
Although use java reflection, it seems good to me. @LuciferYang Thank you.

Reply:
When upgrading to use Scala 3, I think we should be able to avoid using java reflection